docs/PROTOCOL_v21.md

```diff
@@ -0,0 +1,375 @@
+Logs are marked as seen when they are returned by the buffer, to avoid working with logs that have already been seen.
+
+It is used by the log provider to provide unknown logs to the node, and by by the log recoverer to identify known logs during recovery flow.
+
+
+### Log Recoverer
+
+The log recoverer is responsible to ensure that no logs are missed.
+It does that by running a background process for re-scanning of logs and putting the ones we missed into the recovery flow (without checkBlockNum/Hash).
+
+Logs will be considered as missed if they are older than `latestBlock - lookbackBlocks` and has not been performed or successfully checked already (eligible).
+
+While the provider is scanning latest logs, the recoverer is scanning older logs in ascending order, up to `latestBlock - lookbackBlocks`, newer blocks will be under the provider's lookback window.
+
+**Recoverer scanning process**
+
+- The recoverer maintains a `lastRePollBlock` for each upkeep, .i.e. the last block it scanned for that upkeep.
+- Every second, the recoverer will scan logs for a subset of `n=10` upkeeps, where `n/2` upkeeps are randomly chosen and `n/2` upkeeps are chosen by the oldest `lastRePollBlock`.
+- It will start scanning from `lastRePollBlock` on each iteration, and update the block number when it finishes scanning.
+- Logs that are older than 24hr are ignored, therefore `lastRePollBlock` starts at `latestBlock - (24hr block)` in case it was not populated before.
+
+#### Upkeep States
+
+The upkeeps states are used to track the status of log upkeeps (eligible, performed) across the system, to avoid redundant work by the recoverer. Enables to select by (upkeepID, logIdentifier) is used as a key to store the state of a log upkeep.
+
+The state is updated by the coordinator when the upkeep is performed or after positive check (eligible).
+
+The states will be persisted to so the latest state to be restored when the node starts up.
+
+### Plugin
+
+The plugin is performing the following tasks upon OCR3 procedures:
+
+#### Observation
+
+Observation starts with a processing of the previous outcome:
+- Remove agreed upon finalized results from result store
+- For `acceptedSamples` (bound to a trigger)
+    - Remove from metadata store
+    - Enqueue (trigger, upkeepID) into conditional observer
+- For `acceptedRecoveryLogs` (bound to a trigger)
+    - Remove from metadata store
+    - Enqueue (trigger, upkeepID) into recovery observer
+
+Then we do the following for current observation:
+- Query results from result store giving seqNr as pseudoRandom seed and predefined limit. Filter results using coordinator and add them to observation
+- Query from metadata store for sample and recovery instructions using seqNr as pseudoRandom seed within limits, filter using coordinator and add them to observation
+- Query latest block number and hashes from local chain, add them to observation
+
+#### Outcome
+
+- Derive latest blockNumber and blockHash
```

Comment:
This is not done in every outcome. When all the nodes agree that block coordination should be done, then they do it using the block history. The block that achieves quorum is the most recent block/hash that the majority of nodes have in common.